### PR TITLE
backend/feat/marketing-events-addition

### PR DESCRIPTION
**Pull request by RahulGupta0706**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

## Summary by CodeRabbit

- **New Features**
	- Added marketing events tracking for driver signup and first ride completion
	- Implemented notifications for marketing events via Firebase

- **Refactor**
	- Enhanced encryption and data flow handling in driver score event processing
	- Introduced new module for managing marketing event notifications

- **Chores**
	- Updated project configuration to expose new marketing events module

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Registration.hs [hunk 1]
```diff
@@ -399,6 +400,8 @@ createDriverWithDetails req mbBundleVersion mbClientVersion mbClientConfigVersio
   person <- makePerson req transporterConfig mbBundleVersion mbClientVersion mbClientConfigVersion mbDevice mbBackendApp merchantId merchantOpCityId isDashboard Nothing
   void $ QP.create person
   createDriverDetails (person.id) merchantId merchantOpCityId transporterConfig
+  merchantOperatingCity <- CQMOC.findById merchantOpCityId >>= fromMaybeM (MerchantOperatingCityDoesNotExist merchantOpCityId.getId)
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Registration.hs:403]: Fork this

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs [hunk 2]
```diff
@@ -114,6 +116,9 @@ eventPayloadHandler merchantOpCityId DST.OnRideCompletion {..} = do
   -- mbDriverStats <- DSQ.findById (cast driverId) -- always be just because stats will be created at OnNewRideAssigned
   updateDailyStats driverId merchantOpCityId ride fareParameter
   whenJust mbDriverStats $ \driverStats -> do
+    when (driverStats.totalRides == 0) $ do
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:119]: Fork this

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/MarketingEvents.hs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+data MarketingEventsType
+  = NEW_SIGNUP
+  | FIRST_RIDE_COMPLETED
+  deriving (Show)
+
+notifyMarketingEvents :: (EncFlow m r, EsqDBFlow m r, CacheFlow m r) => Id Person.Person -> Id DMOC.MerchantOperatingCity -> ShortId DMerchant.Merchant -> MarketingEventsType -> Maybe VehicleCategory -> DMOC.MerchantOperatingCity -> m ()
+notifyMarketingEvents driverId merchantOpCityId merchantShortId marketingEventsType vehicleCategory city = fork "marketing-events" $ do
+  let events = case vehicleCategory of
+        Just vehicleCat -> (getShortId merchantShortId) <> "driver" <> T.toLower (show marketingEventsType) <> T.toLower (show vehicleCat) <> show city.city
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/MarketingEvents.hs:24]: make full string toLower

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/MarketingEvents.hs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        _ -> (getShortId merchantShortId) <> "driver" <> T.toLower (show marketingEventsType) <> show city.city
+  SQP.findById driverId >>= \case
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/MarketingEvents.hs:26]: Extra query, this is not required. Please remove.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/MarketingEvents.hs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import Storage.Queries.DriverInformation as DI
+import Storage.Queries.Person as SQP
+import Tools.Notifications
+
+data MarketingEventsType
+  = NEW_SIGNUP
+  | FIRST_RIDE_COMPLETED
+  deriving (Show)
+
+notifyMarketingEvents :: (EncFlow m r, EsqDBFlow m r, CacheFlow m r) => Id Person.Person -> Id DMOC.MerchantOperatingCity -> ShortId DMerchant.Merchant -> MarketingEventsType -> Maybe VehicleCategory -> DMOC.MerchantOperatingCity -> m ()
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/MarketingEvents.hs:21]: Pass Person in `notifyMarketingEvents`

**Review by slow-codex — APPROVED** [review]

**Review by khuzema786 — APPROVED** [review]

**Review by khuzema786 — APPROVED** [review]

**Pull request merged**